Update tutorial (#208)
fsspec/kerchunk: docs/source/tutorial.rst

            variable = file.split('/')[-1].split('.')[0]
            month = file.split('/')[2] 
            outf = f'{month}_{variable}.json' #file name to save json to
            with fs2.open(outf, 'wb') as f:
                f.write(ujson.dumps(h5chunks.translate()).encode());

ERA5-pds is located in us-west-2 and so depending on where this computation is taking place the time taken can vary dramatically and is likely to be faster than this.

.. code:: 

    %%time
    for file in flist:
        gen_json(file)

.. parsed-literal::

    CPU times: user 30.4 s, sys: 4.74 s, total: 35.1 s
    Wall time: 14min 44s

The ``.json`` reference files we have generated can now be used to open virtual datasets through xarray. To achieve this it is first necessary to create a mapping of the reference file using ``fsspec``. Here specifying that the reference json is pointing to files stored on AWS.

.. code:: 

    import xarray as xr

    %%time
    fs_ = fsspec.filesystem("reference", fo='01_air_pressure_at_mean_sea_level.json', ref_storage_args={'skip_instance_cache':True},
                           remote_protocol='s3', remote_options={'anon':True})
    m = fs_.get_mapper("")

>     ds = xr.open_dataset(m, engine="zarr", backend_kwargs={'consolidated':False}, chunks={})

Reviewer: You can do all this in one line, but its a bit messy:
```
ds = xr.open_dataset("reference://", engine="zarr", backend_kwargs={
    "consolidate": False,
    "storage_options": {"fo": '01_air_pressure_at_mean_sea_level.json', "remote_protocol": "s3",
        "remote_options": {"anon": True}}
)
```
(the complexity of this line is why people would eventually want to hide such a dataset prescription inside an Intake catalogue).

    import os
    import ujson
    
    so = dict(mode='rb', anon=True, default_fill_cache=False, default_cache_type='first') # args to fs.open()
    # default_fill_cache=False avoids caching data in between file chunks to lowers memory usage.
    
    def gen_json(file):
        with fs.open(file, **so) as infile:
            h5chunks = SingleHdf5ToZarr(infile, file, inline_threshold=300) 
            # inline threshold adjusts the Size below which binary blocks are included directly in the output
            # a higher inline threshold can result in a larger json file but faster loading time
            variable = file.split('/')[-1].split('.')[0]
            month = file.split('/')[2] 
            outf = f'{month}_{variable}.json' #file name to save json to
            with fs2.open(outf, 'wb') as f:
                f.write(ujson.dumps(h5chunks.translate()).encode());

ERA5-pds is located in us-west-2 and so depending on where this computation is taking place the time taken can vary dramatically.

.. code:: 

    %%time
    for file in flist:
        gen_json(file)

.. parsed-literal::

    CPU times: user 30.4 s, sys: 4.74 s, total: 35.1 s
    Wall time: 14min 44s

> The ``.json`` reference files we have generated can now be used to open virtual datasets through xarray. To achieve this it is first necessary to create a mapping of the reference file using ``fsspec``. Here specifying that the reference json is pointing to files stored on AWS.

Author: updated to now read:

The ``.json`` reference files we have generated can now be used to open virtual datasets through xarray or zarr. It is necessary to specify location of the reference ``json`` files, using the ``target_options`` argument, and the source data using the ``remote_options`` and ``remote_protocol`` arguments. Here specifying that the source data is stored on ``AWS S3`` and can be accessed anonymously.

For more complex uses it is also possible to pass in a compiled ``regex`` function which operates on the input file urls to generate a unique variable for each file.

.. code:: 

    import re
    ex = re.compile(r'.*(\d+)_air')
    print(ex.match(json_list[0]).groups()[0])

.. parsed-literal::

    '1'

.. code:: 

    mzz = MultiZarrToZarr(json_list,                
        remote_protocol='s3',
        remote_options={'anon':True},
        coo_map = {'new_dimension':ex},
        concat_dims=['new_dimension'],
        identical_dims = ['lat', 'lon']
    )
    
    d = mzz.translate()

    backend_args = {"consolidated": False, "storage_options": {"fo": d, "remote_protocol": "s3","remote_options": {"anon": True}}}
    print(xr.open_dataset("reference://", engine="zarr", backend_kwargs=backend_args))

.. parsed-literal::

    <xarray.Dataset>
    Dimensions:                         (new_dimension: 2, time0: 744, lat: 721,
                                         lon: 1440)
    Coordinates:
      * lat                             (lat) float32 90.0 89.75 ... -89.75 -90.0
      * lon                             (lon) float32 nan 0.25 0.5 ... 359.5 359.8

>       * new_dimension                   (new_dimension) object '1' '2'

Reviewer: and now this is the result of applying the regex to each input file name

      * time0                           (time0) datetime64[ns] 2020-01-01 ... 202...
    Data variables:
        air_pressure_at_mean_sea_level  (new_dimension, time0, lat, lon) float32 ...
    Attributes:
        institution:  ECMWF
        source:       Reanalysis
        title:        ERA5 forecasts

> Similarly we can map each file to a new variable using the special ``var`` character in coo_map. Here we use the same ``regex`` function but instead map these as new variables.

Reviewer: ```suggestion
Similarly we can map each file to a new variable using the special ``var`` key in coo_map. Here we use the same ``regex`` function but instead map these as new variables.
```

.. code:: 

    mzz = MultiZarrToZarr(json_list,                
        remote_protocol='s3',
        remote_options={'anon':True},
        coo_map = {"var":ex},
        concat_dims=['time0'],
        identical_dims = ['lat', 'lon']
    )
    
    d = mzz.translate()

    backend_args = {"consolidated": False, "storage_options": {"fo": d, "remote_protocol": "s3","remote_options": {"anon": True}}}
    print(xr.open_dataset("reference://", engine="zarr", backend_kwargs=backend_args))

.. parsed-literal::

    <xarray.Dataset>
    Dimensions:  (time0: 1440, lat: 721, lon: 1440)
    Coordinates:
      * lat      (lat) float32 90.0 89.75 89.5 89.25 ... -89.25 -89.5 -89.75 -90.0
      * lon      (lon) float32 nan 0.25 0.5 0.75 1.0 ... 359.0 359.2 359.5 359.8
      * time0    (time0) datetime64[ns] 2020-01-01 ... 2020-02-29T23:00:00
    Data variables:
        1        (time0, lat, lon) float32 ...
        2        (time0, lat, lon) float32 ...
    Attributes:
        institution:  ECMWF
        source:       Reanalysis
        title:        ERA5 forecasts

> Another special character in ``coo_map`` is ``attr:``. This allows the user to access variables from the file attributes.

Reviewer: ```suggestion
Another special character in ``coo_map`` is ``attr:``. This allows the user to access values from the each dataset's global attributes.
```

.. code:: 

    mzz = MultiZarrToZarr(json_list,                
        remote_protocol='s3',
        remote_options={'anon':True},
        coo_map = {"var":"attr:institution"},
        concat_dims=['time0'],
        identical_dims = ['lat', 'lon']
    )
    
    d = mzz.translate()

    backend_args = {"consolidated": False, "storage_options": {"fo": d, "remote_protocol": "s3","remote_options": {"anon": True}}}
    print(xr.open_dataset("reference://", engine="zarr", backend_kwargs=backend_args))

.. parsed-literal::

    <xarray.Dataset>
    Dimensions:  (time0: 1440, lat: 721, lon: 1440)
    Coordinates:
      * lat      (lat) float32 90.0 89.75 89.5 89.25 ... -89.25 -89.5 -89.75 -90.0
      * lon      (lon) float32 nan 0.25 0.5 0.75 1.0 ... 359.0 359.2 359.5 359.8
      * time0    (time0) datetime64[ns] 2020-01-01 ... 2020-02-29T23:00:00
    Data variables:

>         ECMWF    (time0, lat, lon) float32 ...

Reviewer: but both datasets had the save value of the attribute, so is this actually what we expect?

Author: In this case the `json_list` contains two files of consecutive time steps and the same variable

    Attributes:
        institution:  ECMWF
        source:       Reanalysis
        title:        ERA5 forecasts

> The special character ``vattr:{var}:`` allows access to variable attributes. Here renaming the variable to instead use it’s short name. 

Reviewer: ```suggestion
The special value ``vattr:{var}:{attr}`` allows access to variable attributes. Here renaming the variable to instead use its short name. 
```

> There are a number of other special characters for ``coo_map`` documented in the `API reference <https://fsspec.github.io/kerchunk/reference.html#kerchunk.combine.MultiZarrToZarr>`__

Reviewer: This line should follow the example block below

    from kerchunk.hdf import SingleHdf5ToZarr 
    import fsspec

Using fsspec to create a pythonic filesystem, provides a convenient way to manage file urls. 

> The ``SingleHdf5ToZarr`` method takes both an ``h5f`` file and a ``url`` as input. The ``h5f`` file can either be a binary Python file-like object or a url, in which case it will be opened using ``fsspec`` and ``storage_options``. The ``url`` input is not used to open the file and is intended to allow the user to compute the reference files on data before it is uploaded to its final storage location. Thus the ``url`` input should be the url of the final file destination and not the current location. 

Author: I have added this section after out conversation yesterday to make it clear the reference files do not need to be regenerated if the source files move. Although I am not sure how well this comes across here?
